bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/utils/faillog/LogBufferTest.java

```diff
@@ -0,0 +1,315 @@
+    assertArrayEquals(buf1, res.poll());
+    assertArrayEquals(buf2, res.poll());
+    assertNull(res.poll());
+  }
+
+  @Test
+  public void oneEntryLargerThanLimitCanBeAdmitted() throws InterruptedException {
+    final LogBuffer buffer = new LogBuffer(3, true);
+    final byte[] buf1 = new byte[] {0};
+    final byte[] buf2 = new byte[] {1, 2, 3, 4};
+    final byte[] buf3 = new byte[] {5, 6, 7, 8, 9, 10, 11};
+    assertTrue(buffer.append(buf1));
+
+    // `buf2` doesn't get admitted because there is no space.
+    assertFalse(buffer.append(buf2));
+    Queue<byte[]> res = buffer.drain();
+    assertArrayEquals(buf1, res.poll());
+    assertNull(res.poll());
+
+    // `buf3` gets admitted even though it's larger than the total space limit, because it's the
+    // only entry.
+    assertTrue(buffer.append(buf3));
+    res = buffer.drain();
+    assertArrayEquals(buf3, res.poll());
+    assertNull(res.poll());
+  }
+
+  @Test
+  public void overflowBlocks() throws InterruptedException, ExecutionException {
+    final LogBuffer buffer = new LogBuffer(11, false);
+    final byte[] buf1 = new byte[] {0, 1, 2};
+    final byte[] buf2 = new byte[] {3, 4, 5, 6, 7, 8};
+    final byte[] buf3 = new byte[] {9, 10, 11};
+
+    assertTrue(buffer.append(buf1));
+    assertTrue(buffer.append(buf2));
+    // There is no space for `buf3`, so `buffer.append(buf3)` should block, so we run it in a
+    // separate thread.
+    final Future<Void> threadCompletion =
+        runInThread(
+            new Callable<Void>() {
+              @Override
+              public Void call() throws Exception {
+                buffer.append(buf3);
+                return null;
+              }
+            });
+
+    // A reasonably confident check if to verify that the thread is blocked. No false negatives are
+    // possible.
+    assertFutureNotCompleting(threadCompletion);
+
+    Queue<byte[]> res = buffer.drain();
+    assertArrayEquals(buf1, res.poll());
+    assertArrayEquals(buf2, res.poll());
+    assertNull(res.poll());
+
+    // After we've drained, `buf3` should have been finally admitted and thread exited.
+    threadCompletion.get();
+    res = buffer.drain();
+    assertArrayEquals(buf3, res.poll());
+    assertNull(res.poll());
+  }
+
+  @Test
+  public void emptyBufferBlocks() throws InterruptedException, ExecutionException {
+    final LogBuffer buffer = new LogBuffer(11, true);
+    final byte[] buf1 = new byte[] {0, 1, 2};
+
+    final Future<Void> threadCompletion =
+        runInThread(
+            new Callable<Void>() {
+              @Override
+              public Void call() throws Exception {
+                Queue<byte[]> res = buffer.drain();
+                assertArrayEquals(buf1, res.poll());
+                assertNull(res.poll());
+                return null;
+              }
+            });
+
+    // A reasonably confident check to verify that the thread is blocked. No false negatives are
+    // possible.
```

Comment:
I don't think it's that easy.

Internally, the code boils down to:
```
lock.lock();
while (buffers.isEmpty() && !shutdown) {
  notEmpty.await();
}
```
I want to assert that this code won't proceed until the user inserts a buffer.
If I understand you correctly, you'd like to change the code to:
```
lock.lock();
while (buffers.isEmpty() && !shutdown) {
  threadBlocked = true;
  notEmpty.await();
  threadBlocked = false;
}
```
and then, in the test, check that `threadBlocked == true`.

There are a couple of problems with it:
* spurious wake-ups make it more complicated (effectively having to make that flag protected by the same mutex as the condition)
* in my test, I need to somehow wait for `threadBlocked` to become true (so a flag is not enough - we need to wait on it)
* every `Condition` requires a separate flag
* the instrumentation for testing would become more complicated than the code itself with that waiting

I think to make it properly, we'd need to write a wrapper around `Lock` and `Condition`, so that it allowed waiting until a given thread becomes blocked on a condition and expose that from the objects. Honestly, I think that's an overkill with a nontrivial performance penalty.

I think having occasional false positives is not a problem - they would instantly break CI, so they would be fixed promptly - a much smaller problem than false negatives, which constitute flakiness. Also, to make the CI break, usually the false positive would have to be triggered twice in a row - once on your laptop when you're developing code and once in CI - pretty unlikely.